### PR TITLE
Integrate OpenID4VP authentication into OIDC flow

### LINKED ISSUE #116: Import user from credential in OpenID4VP auth flow
### Description

The current implementation assumes the authenticating user is simply matched with users in the Keycloak database, which will not be the case in the upcoming demo, relying on a SAFE IdP.

### PR DESCRIPTION
**Pull request by IngridPuppet**
Closes #116 
Addresses most of, but does not close, https://github.com/ADORSYS-GIS/keycloak-fipa-vc-demo/issues/7
Customization on target theme will be done separately.

### CODE REVIEW

File: services/src/main/java/org/keycloak/protocol/oid4vc/oid4vp/authenticator/SdJwtAuthenticator.java [hunk 1]
```diff
@@ -114,8 +108,28 @@ private SdJwtAuthRequirements getAuthenticationRequirements(AuthenticationFlowCo
...
+        // Import user if not found
+        if (user == null) {
+            // TODO: Improve user import strategy. Extend AbstractIdpAuthenticator?
+            user = context.getSession().users().addUser(context.getRealm(), username);
+            user.setEnabled(true);
+            logger.infof("Imported user '%s' from SD-JWT credential", username);
+        }
+
+        return user;
+    }
```
Comment (forkimenjeckayang) [inline services/src/main/java/org/keycloak/protocol/oid4vc/oid4vp/authenticator/SdJwtAuthenticator.java]: Could the  `recoverAuthenticatingUser` method  include:
- Validation that the username from the credential is not null/empty
- Proper error handling for user creation failures
- Consideration of realm-specific user import policies
- Proper logging of security events for audit purposes

File: services/src/main/java/org/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java [hunk 2]
```diff
@@ -218,15 +219,40 @@ public Response getAuthorizationContextState(@PathParam("transactionId") String
...
+    public ClientModel checkClient(String clientId) {
+        ClientModel client = realm.getClientByClientId(clientId);
+
+        if (client == null) {
+            throw new IllegalArgumentException("Invalid client ID");
+        }
+
+        if (!client.isEnabled()) {
+            throw new IllegalArgumentException("Client is disabled");
+        }
```
Comment (forkimenjeckayang) [inline services/src/main/java/org/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java]: The `checkClient` method only validates client existence and enabled status. 
Could we consider adding:
- Validation that the client is authorized for OID4VP authentication
- Check for client-specific OID4VP configuration
- Proper error messages for different failure scenarios

File: themes/src/main/resources/theme/keycloak.v2+oid4vp/login/resources/js/oid4vp.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    // Authentication was successful
+    submitForm(data.authorization_code);
+  } catch (error) {
+    console.error("Error while polling:", error);
+    // Retrying
+    setTimeout(() => checkAuthStatus(url, period), period);
+  }
+}
```
Comment (forkimenjeckayang) [inline themes/src/main/resources/theme/keycloak.v2+oid4vp/login/resources/js/oid4vp.js]: Some recommendations for this if necessary:
- Add exponential backoff for retry attempts
- Improve error handling for network failures
- Add timeout handling

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/oid4vp/OID4VPUserAuthBeanTest.java
Comment (forkimenjeckayang) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/oid4vp/OID4VPUserAuthBeanTest.java:1]: With some other consideration mentioned earlier, if applied , We may need to include tests for : 
- User import failure scenarios
- Concurrent authentication attempts
- Session timeout handling
- Invalid client configurations
- Network failure scenarios in the polling mechanism

**Review by forkimenjeckayang — CHANGES_REQUESTED** [review]
In addition to other comments , could this PR include include:
- Configuration documentation for the OID4VP authentication flow and setup instructions for administrators

About the TODO mentioned in `SdJwtAuthenticator` I guess that will be handled in another PR later ?

**Comment by IngridPuppet** [comment]
> In addition to other comments , could this PR include include:
> 
> * Configuration documentation for the OID4VP authentication flow and setup instructions for administrators

Thank you for your review. There is not really not much to do, the feature is intended to work out of the box. Yes, you need to enable the right feature flags and login theme. The former was already mentioned. I amended the docs to mention the latter.

> About the TODO mentioned in `SdJwtAuthenticator` I guess that will be handled in another PR later ?

Let's say yes. But that will technically not be necessary with the new approach around identity providers we're heading toward.

File: services/src/main/java/org/keycloak/protocol/oid4vc/oid4vp/authenticator/SdJwtAuthenticator.java (hunk 1 above)
Comment (IngridPuppet) [reply services/src/main/java/org/keycloak/protocol/oid4vc/oid4vp/authenticator/SdJwtAuthenticator.java]: > Validation that the username from the credential is not null/empty

Not really necessary because that can't happen. See the reason in the implementation of `readUsernameFromCredential`. If for some weird reason it does, the flow will stop there.

> Proper error handling for user creation failures.

Thank you for catching the possible NPE. I added back the code I previously removed for this, namely `failDenyingAuthenticatingUser`.

> Consideration of realm-specific user import policies

Let's say this is part of the TODO. It would be overkill for the moment because this manual logic is very not likely to stay.

> Proper logging of security events for audit purposes

I get how important it might be, but we may skip that for now to meet our deadline.

File: services/src/main/java/org/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java (hunk 2 above)
Comment (IngridPuppet) [reply services/src/main/java/org/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java]: > * Validation that the client is authorized for OID4VP authentication
> * Check for client-specific OID4VP configuration

In the same way other authentication flows can be enabled or disabled per client? Interesting, but let's make it a TODO. We really don't want to complexify the support of the feature for now. As previously mentioned, we're even considering a completely different approach around identity providers, and that will not be necessary anymore.

Comment (IngridPuppet) [reply services/src/main/java/org/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java]: > Proper error messages for different failure scenarios

I added one for inexistent clients. Please could you check?

File: themes/src/main/resources/theme/keycloak.v2+oid4vp/login/resources/js/oid4vp.js (hunk 3 above)
Comment (IngridPuppet) [reply themes/src/main/resources/theme/keycloak.v2+oid4vp/login/resources/js/oid4vp.js]: Considering this is not the final template, I think it is preferable to invest more time for this into the new one. Did you have some suggestions for improving error handling regarding network failures? I updated the script to stop polling after a fixed number of network failures. Exponential backoff can be problematic because we want the page to be as responsive as possible. Note that in case of no network failure, the authentication session naturally expires and one poll will end in 404.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/oid4vp/OID4VPUserAuthBeanTest.java
Comment (IngridPuppet) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/oid4vp/OID4VPUserAuthBeanTest.java:1]: These are good test suggestions. Considering we need to ship the feature faster, I believe it is not appropriate to spend time in crafting complicated tests for extreme edge cases.

> User import failure scenarios

A recovery path was implemented but that can be very cumbersome to demonstrate in tests. Logs will help in case.

> Concurrent authentication attempts

We shouldn't worry too much given that Keycloak spins up endpoints with this in mind. Private class data are typically not shared across sessions. Also, different authentication sessions are created for this flow.

> Session timeout handling

Already covered as we reuse an existing stack for managing these auth sessions.

> Invalid client configurations

One was added.

> Network failure scenarios in the polling mechanism

Cumbersome to set up. It is really not the right time for that.

File: services/src/main/java/org/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java [hunk 4]
```diff
@@ -86,29 +88,29 @@ public Response preflight() {
...
-                .createAuthorizationRequest(authSession, authReqs);
+        try {
+            checkClient(clientId);
+        } catch (IllegalArgumentException e) {
+            throw new BadRequestException(errorResponse(
+                    Response.Status.BAD_REQUEST,
+                    OAuthErrorException.INVALID_CLIENT,
+                    "Invalid client ID. Maybe unknown or disabled"
+            ), e);
+        }
```
Comment (Awambeng) [inline services/src/main/java/org/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java]: The error message "Maybe unknown or disabled" could be more specific. Consider checking the client status separately to provide more precise error messages.

File: services/src/main/java/org/keycloak/forms/login/freemarker/model/OID4VPUserAuthBean.java [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+                    Map.of(EncodeHintType.MARGIN, 0)
+            );
+
+            ByteArrayOutputStream bos = new ByteArrayOutputStream();
+            MatrixToImageWriter.writeToStream(bitMatrix, QR_CODE_IMAGE_FORMAT, bos);
+
+            byte[] pngBytes = bos.toByteArray();
+            String base64 = Base64.getEncoder().encodeToString(pngBytes);
+
+            return String.format("data:image/%s;base64,%s", QR_CODE_IMAGE_FORMAT, base64);
```
Comment (Awambeng) [inline services/src/main/java/org/keycloak/forms/login/freemarker/model/OID4VPUserAuthBean.java:172]: Consider adding input validation for the data parameter to prevent potential issues with very large QR codes.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/oid4vp/OID4VPUserAuthEndpointTest.java [hunk 6]
```diff
@@ -721,27 +765,82 @@ private String getOid4vpEndpoint(String route) {
...
-        boolean shouldBase64EncodeVpToken;
-        String overridePresentationDefinitionId;
-        Descriptor.Format overrideDescriptorFormat;
-        String overrideDescriptorPath;
+        private String testUser = TEST_USER;
+        private boolean shouldBase64EncodeVpToken;
+        private boolean shouldRetrieveAccessToken = true;
+        private String overridePresentationDefinitionId;
+        private Descriptor.Format overrideDescriptorFormat;
+        private String overrideDescriptorPath;
```
Comment (Awambeng) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/oid4vp/OID4VPUserAuthEndpointTest.java:829]: This is a well-designed builder pattern implementation. Consider adding validation in the setter methods to ensure valid values are provided.

File: themes/src/main/resources/theme/keycloak.v2+oid4vp/login/resources/js/oid4vp.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+    // Authentication was successful
+    submitForm(data.authorization_code);
+  } catch (error) {
+    console.error("Error while polling:", error);
+    // Retrying
+    setTimeout(() => checkAuthStatus(
+      url, period, currentRetry + 1, maxRetries
+    ), period);
+  }
```
Comment (Awambeng) [inline themes/src/main/resources/theme/keycloak.v2+oid4vp/login/resources/js/oid4vp.js:46]: Consider adding a maximum retry limit to prevent infinite polling in case of persistent errors. Also, the error handling could be more specific for different types of errors.

**Review by Awambeng — CHANGES_REQUESTED** [review]

File: services/src/main/java/org/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java (hunk 4 above)
Comment (IngridPuppet) [reply services/src/main/java/org/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java]: Thank you for your review. Considering the helper evolves and the client could be invalid for many more reasons, it is actually not sustainable to recheck every failing condition. As per Keycloak's recent commitment to not leak uncontrolled error messages to the user, I've replaced the message here with a generic one. The specific reason will be available in logs.

File: services/src/main/java/org/keycloak/forms/login/freemarker/model/OID4VPUserAuthBean.java (hunk 5 above)
Comment (IngridPuppet) [reply services/src/main/java/org/keycloak/forms/login/freemarker/model/OID4VPUserAuthBean.java:172]: The QR code generator module knows best how to run this validation. If any issue, it throws an exception, which is caught and handled here.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/oid4vp/OID4VPUserAuthEndpointTest.java (hunk 6 above)
Comment (IngridPuppet) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/oid4vp/OID4VPUserAuthEndpointTest.java:829]: Thank you for your suggestion. Considering this is a test construct, this is not really worth the time it would take.

File: themes/src/main/resources/theme/keycloak.v2+oid4vp/login/resources/js/oid4vp.js (hunk 7 above)
Comment (IngridPuppet) [reply themes/src/main/resources/theme/keycloak.v2+oid4vp/login/resources/js/oid4vp.js:46]: > Consider adding a maximum retry limit to prevent infinite polling in case of persistent errors.

There is a `maxRetries` logic implemented. Did you check the file before I addressed Forkim's comments?

> Also, the error handling could be more specific for different types of errors.

May I know your mind? A `fetch` may fail for a ton of reasons. From where would you start and what would be the added benefit?

Comment (Awambeng) [reply themes/src/main/resources/theme/keycloak.v2+oid4vp/login/resources/js/oid4vp.js:46]: > > Consider adding a maximum retry limit to prevent infinite polling in case of persistent errors.
> 
> There is a `maxRetries` logic implemented. Did you check the file before I addressed Forkim's comments?
> 
> > Also, the error handling could be more specific for different types of errors.
> 
> May I know your mind? A `fetch` may fail for a ton of reasons. From where would you start and what would be the added benefit?

I checked again and confirmed that the `maxRetries` logic is indeed implemented. I also agree with you that a fetch can fail for numerous reasons, and trying to handle all of them would be complex. That said, the current implementation looks solid. Thank you!

**Review by Awambeng — COMMENTED** [review]
@IngridPuppet thank you for considering my comments, and apologies for causing extra work with the conflicts. I had to rebase our `datev/develop` branch onto main.

**Comment by IngridPuppet** [comment]
> @IngridPuppet thank you for considering my comments, and apologies for causing extra work with the conflicts. I had to rebase our `datev/develop` branch onto main.

Fortunately, I faced no conflicts updating the PR.

**Review by Awambeng — APPROVED** [review]

File: themes/src/main/resources/theme/keycloak.v2+oid4vp/login/resources/js/oid4vp.js (hunk 3 above)
Comment (forkimenjeckayang) [reply themes/src/main/resources/theme/keycloak.v2+oid4vp/login/resources/js/oid4vp.js]: SInce we already have
- Retry limit system 
- Basic error handling 
- Status differentiation
- Timeout protection
- Error logging we can work with that for now. As we have a deadline to meet.

**Review by forkimenjeckayang — APPROVED** [review]
